### ISSUE TITLE
Unable to run tests in parallel

### ISSUE THREAD

**Issue by tjheslin1** [manub/scalatest-embedded-kafka#101]
Due to the singleton, mutable `broker` ([see here](https://github.com/manub/scalatest-embedded-kafka/blob/master/embedded-kafka/src/main/scala/net/manub/embeddedkafka/EmbeddedKafka.scala#L47)). It's not possible to run tests in parallel, even when randomising ports.

If more than one test using `EmbededKafka` runs at the same time I consistently get:
```
FATAL [Replica Manager on Broker 0]: Error writing to highwatermark file:  (kafka.server.ReplicaManager:118)
java.io.FileNotFoundException: /var/folders/5l/z1qx5k790f9_h1wtxk8mj1k4pgw6zm/T/kafka-logs4978987266055584410.tmp/replication-offset-checkpoint.tmp (No such file or directory)
        at java.io.FileOutputStream.open0(Native Method)
        at java.io.FileOutputStream.open(FileOutputStream.java:270)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
        at kafka.server.checkpoints.CheckpointFile.write(CheckpointFile.scala:43)
        at kafka.server.checkpoints.OffsetCheckpointFile.write(OffsetCheckpointFile.scala:58)
        at kafka.server.ReplicaManager.$anonfun$checkpointHighWatermarks$5(ReplicaManager.scala:1118)
        at kafka.server.ReplicaManager.$anonfun$checkpointHighWatermarks$5$adapted(ReplicaManager.scala:1115)
        at scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:789)
        at scala.collection.immutable.Map$Map1.foreach(Map.scala:120)
        at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:788)
        at kafka.server.ReplicaManager.checkpointHighWatermarks(ReplicaManager.scala:1115)
        at kafka.server.ReplicaManager.$anonfun$startHighWaterMarksCheckPointThread$1(ReplicaManager.scala:211)
        at kafka.utils.KafkaScheduler.$anonfun$schedule$2(KafkaScheduler.scala:110)
        at kafka.utils.CoreUtils$$anon$1.run(CoreUtils.scala:57)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

I think it would make sense that `start()` returned a new instance of `KafkaServer` each time it is called.

**Comment by manub** [comment]
Hi @tjheslin1 and sorry for the very late response. I'll look into this but I have very little spare time in my hands right now - if you could submit a PR I'd more than happily review it quickly otherwise this will have to wait a bit I'm afraid.

**Comment by Thomas-Heslin** [comment]
@manub no worries. I'll have a look into raising a PR soon 👍

**Comment by mwz** [comment]
Hi @Thomas-Heslin, have you been able to make any progress on this?

**Comment by tjheslin1** [comment]
@mwz I haven't unfortunately. I will get round to it after I've addressed an Issue on my own project. Feel free to take it on if you need it sooner.

**Comment by tjheslin1** [comment]
@mwz @manub I am working on this now 👍 Should have a PR soon.

**Comment by mwz** [comment]
Sweet!

**Comment by shashidesai** [comment]
@tjheslin1 - adding to the request. I was just wondering will your fix/solution take care if we are using `runStreams` - so that we can use random available ports ? 
reference - https://github.com/manub/scalatest-embedded-kafka/blob/v0.15.1/kafka-streams/src/test/scala/net/manub/embeddedkafka/streams/ExampleKafkaStreamsSpec.scala#L30

**Comment by tjheslin1** [comment]
Hi @shashidesai I hadn't considered the Kafka Streams side of the project but I will look into your problem too. :D

**Comment by tjheslin1** [comment]
Hi @shashidesai. Looking into your request now. I can see that `runStreams` calls the function `withRunningKafka` which starts a Kafka instance using an implicit `EmbeddedKafkaConfig`.

The Kafka instance created by `withRunningKafka`, from what I can see, is correctly shutdown and shouldn't affect another running instance. With that said, do you have an example of your issue?

What I can do, for consistency, is change the call `withRunningKafka` makes to the function I updated, which starts up Kafka and manages its life-cycle, see [startKafka](https://github.com/tjheslin1/scalatest-embedded-kafka/blob/master/embedded-kafka/src/main/scala/net/manub/embeddedkafka/EmbeddedKafka.scala#L69).

As I don't fully understand your problem yet I wouldn't be surprised if my changes, as they stand, are unable to resolve your issue. I would like to help however!

I would suggest testing out your scenario, explicitly passing an `EmbeddedKafkaConfig`  for each call to `withRunningKafka`. If this fixes your issue I feel the solution would lie in streamlining how this function works. Alternatively you could set the `kafkaPort` in the config to `0`which I believe would randomie the kafka port of each instance.

**Comment by shashidesai** [comment]
Thanks @tjheslin1 
we are using the runStreams like this example  - https://github.com/manub/scalatest-embedded-kafka/blob/v0.15.1/kafka-streams/src/test/scala/net/manub/embeddedkafka/streams/ExampleKafkaStreamsSpec.scala#L30

So I am using this way to run kstream tests, but when I initialize `EmbeddedKafkaConfig`, it uses a port which I want to be dynamic. I don't use `withRunningKafka` in the `runStreams`. if you see the example above in the link provided, it doesn't use `withRunningKafka`.  But when you have multiple github PRs running, only one can run as the port provided in the `EmbeddedKafkaConfig` is constant - so that's where i wanted to know how to get random available ports to run `runStreams`.

does that clarify the issue I am having ?

**Comment by tjheslin1** [comment]
@shashidesai I think I better understand your problem. Can you confirm that setting the port to `0` doesn't choose a randomly available port? 

Looking at `runStreams`, called in the test you linked, I can see that it does call `withRunningKafka`, internally, see here: 
https://github.com/manub/scalatest-embedded-kafka/blob/master/kafka-streams/src/main/scala/net/manub/embeddedkafka/streams/EmbeddedKafkaStreams.scala#L45.  
I was thinking of updating this function, however it would lead to a much bigger set of changes.

I believe the solution to your problem would be in kafka choosing a random available port. This would work in the scenario you've outlined: multiple instances of your app running in github PRs. For this I would I raise a separate issue if the functionality isn't already there in the `KafkaServer` in the Kafka project.

**Comment by tjheslin1** [comment]
Hi @shashidesai. I see. Thanks for trying! I don't believe my change will have an affect as it deals with managing kafka (and zookeeper) instances running in the same app, whereas with mutliple PRs these will be still separate.

**Comment by shashidesai** [comment]
got it. thanks @tjheslin1 . I can make a separate issue for that issue.

**Comment by tjheslin1** [comment]
Cool. I'll be happy to help where I can!
